### PR TITLE
Openidconnect

### PR DESCRIPTION
**Pull request by sigmunau**
This pull request adds basic OpenID Connect authentication support based on the current google authentication. In fact the current google authentication can potentially be dropped in favor of this, since google is also an OpenID Connect provider. 

Here is a list of certified openid connect implementations. This should give an idea about which services that could be used for authentication with this feature: http://openid.net/certification/

### CODE REVIEW

**Comment by chancez** [comment]
This is amazing! I would love if this could make it into the next release. I'm hoping to use this with [coreos/dex](/coreos/dex) as our provider, and this would remove the need for the [bitly/oauth2_proxy](/bitly/oauth2_proxy) im running now.

**Comment by torkelo** [comment]
sorry for delay in getting to this PR, it should be fine to go into core. But will have to review it and maybe do some refactoring. We are close to release 3.1 but will try to get to this in 3.2

**Comment by chancez** [comment]
On second glance, this seems to be doing the implicit flow, which makes this quite similar to another PR I've seen for generic oauth. It would be more ideal to probably wait until there's a full implementation that does key management, and JWT verification.

**Comment by sigmunau** [comment]
It does indeed do implicit flow and is quite similar to the generic oauth pull request which I found only after I had written this. I still think this adds value as it does support reading most of the configuration options directly from the providers configuration end point, and it supports the openid connect defined userinfo format. The generic oauth pull request assumes that the oauth provider has a userinfo endpoint that uses the format of github iirc.

**Comment by Starefossen** [comment]
Any progress on this one? This would make keycloak integration much easier 😄

**Comment by sigmunau** [comment]
We are running off this branch in our shop and it works well for us. It turned out there were another pull request for generic OAuth, which were further along the merge process than this. These to pull requests would most certainly conflict and probably need some refactoring to reduce code duplication. I see the other request is now merged so once I get some time to do it I should probably try to rebase and refactor the code on top of that

**Comment by torkelo** [comment]
closing this as it's drifted a bit too much from master. Grafana backend auth options have expanded quite a bit and we are reaching a point where I think more needs to be handled via backend auth plugins (a concept that needs to be invented & implemented):

https://github.com/grafana/grafana/issues/6841

**Comment by sigmunau** [comment]
Backend auth plugins sounds like the way to go. Waiting for it to be invented and implemented :)

**Comment by ashb** [comment]
If I attempted a PR that added in support for OpenID connect would it be accepted or are we waiting for the backend plugin feature to be implemented?

My goal is to use Azure ActiveDirectory for auth, and it is tantalizingly close with no code changes (by just pretending it's Google since enough is configurable). The only difference is that my AD doesn't give userinfo in the same way -- probably , so I'd need to support looking at the `id_token` return field which is a JWT -- https://github.com/square/go-jose provides all the code we need to validate and decode the JWT token.

In short: worth my time or not?

It is probably possible to make the existing Google auth backend become a simple layer on top of this openid connect one.

**Comment by ashb** [comment]
Actually, there's only one thing I _need_ right now is change the field name that the email comes in in `pkg/social/google_oauth.go`(and I suspect this is because the AD is synced from a corporate managed AD which has odd field names) so I could possibly get away with just making the `SocialGoogle.UserInfo` field configurable.
